src/it/scala/io/iohk/ethereum/db/DataSourceIntegrationTestBehavior.scala

```diff
@@ -0,0 +1,157 @@
+import io.iohk.ethereum.db.dataSource.DataSource
+import org.scalatest.FlatSpec
+import org.scalatest.prop.PropertyChecks
+
+trait DataSourceIntegrationTestBehavior
+  extends PropertyChecks
+    with ObjectGenerators {
+
+  this: FlatSpec =>
+
+  val KeySizeWithoutPrefix: Int = 32
+  val KeySize: Int = KeySizeWithoutPrefix + 1
+  //Hash size + prefix
+  val KeyNumberLimit: Int = 40
+  val OtherNamespace: IndexedSeq[Byte] = IndexedSeq[Byte]('e'.toByte)
+
+  val MaxIncreaseInLength = 10
+
+
+  def withDir(testCode: String => Any): Unit = {
+    val path = Files.createTempDirectory("testdb").getFileName.toString
+    try {
+      testCode(path)
+    } finally {
+      val dir = new File(path)
+      assert(!dir.exists() || dir.delete(), "File deletion failed")
+    }
+  }
+
+  def updateInSeparateCalls(dataSource: DataSource, toUpsert: Seq[(ByteString, ByteString)]): DataSource = {
+    toUpsert.foldLeft(dataSource) { case (recDB, keyValuePair) =>
+      recDB.update(OtherNamespace, Seq(), Seq(keyValuePair))
+    }
+  }
+
+  // scalastyle:off
+  def dataSource(createDataSource: => String => DataSource): Unit = {
+    it should "be able to insert keys in separate updates" in {
+      forAll(seqByteStringOfNItemsGen(KeySizeWithoutPrefix)) { unFilteredKeyList: Seq[ByteString] =>
+        withDir { path =>
+          val keyList = unFilteredKeyList.take(KeyNumberLimit)
+          val db = updateInSeparateCalls(
+            dataSource = createDataSource(path),
+            toUpsert = keyList.zip(keyList)
+          )
+          keyList.foreach { key => assert(db.get(OtherNamespace, key).contains(key)) }
+
+          db.destroy()
+        }
+      }
+    }
+
+    it should "be able to insert keys in a single update" in {
+      forAll(seqByteStringOfNItemsGen(KeySizeWithoutPrefix)) { unFilteredKeyList: Seq[ByteString] =>
+        withDir { path =>
+          val keyList = unFilteredKeyList.take(KeyNumberLimit)
+          val db = createDataSource(path).update(OtherNamespace, Seq(), keyList.zip(keyList))
+
+          keyList.foreach { key => assert(db.get(OtherNamespace, key).contains(key)) }
+
+          db.destroy()
+        }
+      }
+    }
+
+    it should "be able to update keys in separate updates" in {
+      forAll(seqByteStringOfNItemsGen(KeySizeWithoutPrefix)) { unFilteredKeyList: Seq[ByteString] =>
+        withDir { path =>
+          val keyList = unFilteredKeyList.take(KeyNumberLimit)
+          val db = createDataSource(path).update(OtherNamespace, Seq(), keyList.zip(keyList))
+
+          val keyListWithExtraByte = keyList.map(1.toByte +: _)
+          val dbAfterUpdate = updateInSeparateCalls(db, keyList.zip(keyListWithExtraByte))
+
+          keyList.zip(keyListWithExtraByte).foreach { case (key, value) =>
+            assert(dbAfterUpdate.get(OtherNamespace, key).contains(value))
+          }
+
+          dbAfterUpdate.destroy()
+        }
+      }
+    }
+
+    it should "be able to update keys in a single update" in {
```

Comment:
@adamsmo the implementation we are using of levelDB seems to be thread safe [1](https://github.com/dain/leveldb/blob/master/leveldb/src/main/java/org/iq80/leveldb/impl/DbImpl.java#L682) and [2](https://github.com/dain/leveldb/blob/master/leveldb/src/main/java/org/iq80/leveldb/impl/DbImpl.java#L584) (havent checked jni one but I guess that too).

That being said it might be interesting to have this kind of tests, were you thinking these ones in terms of performance or thread safety?